### DISCUSSION TITLE
Recursive record contracts

### DISCUSSION THREAD

**Discussion by suimong** [tweag/nickel#1528] [Q&A] [answered]
To illustrate my use case and the meaning of "recursive record contracts", here's the example: I'm building a contract for GitLab CI/CD, i.e. the `.gitlab-ci.yml` file ([ref](https://docs.gitlab.com/ee/ci/yaml/index.html)).  In essence the yaml file can be modeled by a dictionary contract `{_: JobDefinition}` where the keys are the job names. Within `JobDefinition` there's a `trigger` keyword which defines a series of jobs that will trigger a downstream pipeline. So the `JobDefinition` should look like this:

```
let JobDefinition = {
  image | String,
  trigger | optional | Array JobDefinition,
# other fields omitted
}
```

It's obvious that an infinite loop will occur without the "optional" metadata. On the other hand, with the "optional" metadata, the contract very much resembles a typical recursive linked list definition in OCaml and I'm tempted to think that my contract definition should work, but currently it does not: Nickel reports that the `JobDefinition` attached to `trigger` is  "unbound".

Searching through the repo, "recursive" is more often mentioned in the context of "recursive records" where the "fields" within the record are the subject of recursion, in contrast to this case where the whole contract is the subject of recursion. 

I wonder if my example is somewhat expressible in Nickel?  Or is there plan for Nickel to support this use case?

**Comment by vkleen** [comment, marked as answer]
Nickel has a recursive `let` binding:
```
let rec JobDefinition = {
  image | String,
  trigger | optional | Array JobDefinition,
  # other fields omitted
} in ...
```

**Reply by suimong** [reply]
Ah... my bad. Thanks @vkleen for the quick response, as always. Btw, is there plan to improve the documentation site, like adding search functionality? I definitely would like to help.

**Reply by vkleen** [reply]
I don't think we have existing plans for that but it sounds like a great idea! The website's repository is [https://github.com/tweag/nickel-lang.org](https://github.com/tweag/nickel-lang.org). If you feel like contributing there, even just submitting issues, that would be awesome.
